source/extensions/filters/udp/dns_filter/dns_parser.cc

```diff
@@ -0,0 +1,305 @@
+
+  do {
+    // Read the name segment length or flag;
+    c = buffer->peekBEInt<unsigned char>(*name_offset);
+    *name_offset += sizeof(unsigned char);
+    *available_bytes -= sizeof(unsigned char);
+
+    if (c == 0xc0) {
+      // This is a compressed response. Get the offset in the query record where the domain name
+      // begins. This is done to reduce the name duplication in DNS answer buffers.
+      c = buffer->peekBEInt<unsigned char>(*name_offset);
+
+      // We will restart the loop from this offset and read until we encounter a null byte
+      // signifying the end of the name
+      *name_offset = static_cast<uint64_t>(c);
+
+      continue;
+
+    } else if (c == 0x00) {
+      // We've reached the end of the query.
+      ENVOY_LOG(trace, "End of name: [{}] {}", name_ss.str(), *name_offset);
+      break;
+    }
+
+    const uint64_t segment_length = static_cast<uint64_t>(c);
+
+    // Verify that we have enough data to read the segment length
+    if (segment_length > *available_bytes) {
+      ENVOY_LOG(error,
+                "Insufficient data in buffer for name segment. "
+                "available bytes: {}  segment length: {}",
+                *available_bytes, segment_length);
+      return EMPTY_STRING;
+    }
+
+    // Add the name separator if we have already accumulated name data
+    if (name_ss.tellp()) {
+      name_ss << '.';
+    }
+
+    *available_bytes -= segment_length;
+
+    // The value read is a name segment length
+    for (uint64_t index = 0; index < segment_length; index++) {
+      c = buffer->peekBEInt<unsigned char>(*name_offset);
+      *name_offset += sizeof(unsigned char);
+      name_ss << c;
+    }
+
+  } while (c != 0x00);
+
+  const std::string name = name_ss.str();
+
+  return name;
+}
+
+DnsQueryRecordPtr DnsMessageParser::parseDnsQueryRecord(const Buffer::InstancePtr& buffer,
+                                                        uint64_t* offset) {
+  uint64_t name_offset = *offset;
+  uint64_t available_bytes = buffer->length() - name_offset;
+
+  const std::string record_name = parseDnsNameRecord(buffer, &available_bytes, &name_offset);
+  if (record_name.empty()) {
+    ENVOY_LOG(error, "Unable to parse name record from buffer");
+    return nullptr;
+  }
+
+  if (available_bytes < 2 * sizeof(uint16_t)) {
+    ENVOY_LOG(error, "Insufficient data in buffer to read query record type and class. ");
+    return nullptr;
+  }
+
+  // Read the record type (A or AAAA)
+  uint16_t record_type;
+  record_type = buffer->peekBEInt<uint16_t>(name_offset);
+  name_offset += sizeof(record_type);
+
+  // Read the record class. This value is almost always 1 for internet address records
+  uint16_t record_class;
+  record_class = buffer->peekBEInt<uint16_t>(name_offset);
+  name_offset += sizeof(record_class);
+
+  // This is shared because we use the query from a list when building the response.
+  // Using a shared pointer avoids duplicating this data in the asynchronous resolution path
```

Comment:
I didn't follow why this is shared_ptr

Reply:
In the async resolution path, the query object is what I use to associate the response to the query.   If I don't use a shared pointer, I would have to clone the query when doing the async resolution, since that same object is used to construct the response to the client.

Reply:
It's not actually shared?

Reply:
Fixed